src/java/org/apache/cassandra/index/sai/disk/hnsw/OnDiskOrdinalsMap.java

```diff
@@ -87,6 +94,10 @@ public int[] getSegmentRowIdsMatching(int vectorOrdinal) throws IOException
         {
             Preconditions.checkArgument(vectorOrdinal < size, "vectorOrdinal %s is out of bounds %s", vectorOrdinal, size);
 
+            if (rowIdsMatchOrdinals) {
```

Comment:
I would prefer to handle the special case as a simpler View class, which will save the overhead of unnecessary createReader calls.  then the if statement moves into the getView call